File: 2023/typescript/src/day02.ts
import * as fs from "fs";
import _ from "lodash";

day02();

function day02(): void {
  const lines: string[] = fs
    .readFileSync("data/day02-input.txt", { encoding: "utf8" })
    .trim()
    .split("\n");
  console.log(_.sum(lines.map((l, i) => (isGamePossible(parseGame(l)) ? i + 1 : 0))));
  console.log(_.sum(lines.map((l) => cubeSetPower(minSuperset(parseGame(l))))));
}

interface CubeSet {
  red: number;
  green: number;
  blue: number;
}

function newCubeSet({
  red = 0,
  green = 0,
  blue = 0,
}: {
  red?: number;
  green?: number;
  blue?: number;
}): CubeSet {
  return { red, green, blue };
}

function parseGame(line: string): CubeSet[] {
  return line.split(": ")[1].split("; ").map(parseCubeSet);
}

function parseCubeSet(str: string): CubeSet {
  return str.split(", ").map(parseCubeColorPick).reduce(concatCubeSets);
}

function parseCubeColorPick(str: string): CubeSet {
  const [numStr, color] = str.split(" ");
  const num = parseInt(numStr);
  switch (color) {
    case "red":
      return newCubeSet({ red: num });
    case "green":
      return newCubeSet({ green: num });
    case "blue":
      return newCubeSet({ blue: num });
    default:
      throw new Error("Invalid cube color");
  }
}

function concatCubeSets(cs1: CubeSet, cs2: CubeSet): CubeSet {
  return {
    red: cs1.red + cs2.red,
    green: cs1.green + cs2.green,
    blue: cs1.blue + cs2.blue,
  };
}

function isSubsetOf(cs1: CubeSet, cs2: CubeSet): boolean {
  return cs1.red <= cs2.red && cs1.green <= cs2.green && cs1.blue <= cs2.blue;
}

function isGamePossible(cubeSets: CubeSet[]): boolean {
  return cubeSets.every((cs) => isSubsetOf(cs, { red: 12, green: 13, blue: 14 }));
}

function minSuperset(cubeSets: CubeSet[]): CubeSet {
  return cubeSets.reduce((cb1, cb2) => ({
    red: Math.max(cb1.red, cb2.red),
    green: Math.max(cb1.green, cb2.green),
    blue: Math.max(cb1.blue, cb2.blue),
  }));
}

function cubeSetPower(cs: CubeSet): number {
  return cs.red * cs.green * cs.blue;
}

File: 2023/haskell/src/Day02.hs
module Day02 where

import Data.List.Split (splitOn)
import Paths_aoc2023 (getDataFileName)

-- Game 1: 3 blue, 4 red; 1 red, 2 green, 6 blue; 2 green
-- Game 2: 1 blue, 2 green; 3 green, 4 blue, 1 red; 1 green, 1 blue
-- Game 3: 8 green, 6 blue, 20 red; 5 blue, 4 red, 13 green; 5 green, 1 red
-- Game 4: 1 green, 3 red, 6 blue; 3 green, 6 red; 3 green, 15 blue, 14 red
-- Game 5: 6 red, 1 blue, 3 green; 2 blue, 1 red, 2 green

day02 :: IO ()
day02 = do
  inputLines <- lines <$> (getDataFileName "day02-input.txt" >>= readFile)
  print $ sum $ map fst $ filter (isGamePossible . snd) $ zip [1 :: Int ..] $ parseGame <$> inputLines

data CubeSet = CubeSet {red :: Int, green :: Int, blue :: Int}
  deriving (Show)

instance Monoid CubeSet where
  mempty = CubeSet 0 0 0

instance Semigroup CubeSet where
  (CubeSet a b c) <> (CubeSet a' b' c') = CubeSet (a + a') (b + b') (c + c')

parseGame :: String -> [CubeSet]
parseGame = (parseCubeSet <$>) . splitOn "; " . (!! 1) . splitOn ": "
  where
    parseCubeSet :: String -> CubeSet
    parseCubeSet = mconcat . (parseCubeColorPick <$>) . splitOn ", "

    parseCubeColorPick :: String -> CubeSet
    parseCubeColorPick str = case splitOn " " str of
      [num, color]
        | color == "red" -> mempty {red = read num}
        | color == "green" -> mempty {green = read num}
        | color == "blue" -> mempty {blue = read num}
      _ -> error "invalid cube color pick"

isSubsetOf :: CubeSet -> CubeSet -> Bool
isSubsetOf (CubeSet a b c) (CubeSet a' b' c') = a <= a' && b <= b' && c <= c'

isGamePossible :: [CubeSet] -> Bool
isGamePossible = all (`isSubsetOf` maxSet)
  where
    maxSet = CubeSet {red = 12, green = 13, blue = 14}
